Allowing the usage of input image of type float

jbvimort opened the pull request.

jbvimort committed:
ENH: improvement of the examples in python

jbvimort committed:
ENH: Allowing the usage of float as input image

blowekamp commented on include/itkCoocurrenceTextureFeaturesImageFilter.h:
```diff
@@ -198,8 +198,8 @@ class ITK_TEMPLATE_EXPORT CoocurrenceTextureFeaturesImageFilter
-  itkConceptMacro( InputPixelTypeCheck,
-                   ( Concept::IsInteger< typename InputImageType::PixelType>) );
+//  itkConceptMacro( InputPixelTypeCheck,
+//                   ( Concept::IsInteger< typename InputImageType::PixelType>) );
```
Please just remove the code.

blowekamp commented on include/itkCoocurrenceTextureFeaturesImageFilter.hxx:
```diff
@@ -236,7 +236,7 @@ CoocurrenceTextureFeaturesImageFilter<TInputImage, TOutputImage, TMaskImage>
           // Increase the corresponding bin in the histogram
           ++totalNumberOfFreq;
-          ++hist[currentInNeighborhoodPixelIntensity][pixelIntensity];
+          ++hist[(int)currentInNeighborhoodPixelIntensity][(int)pixelIntensity];
```
I'd recommend adding a protected class typedef "HistogramIndexType" in stead of using these int types through out the code.

blowekamp replied:
floating points are just being casted, this may not give consistent across platform behavior. Consider using a rounding method from itk::Math.

What happened to floating point errors (Nan,Inf)?

Are out of bounds number guarded?

blowekamp commented on include/itkRunLengthTextureFeaturesImageFilter.h:
```diff
@@ -222,8 +222,8 @@ class ITK_TEMPLATE_EXPORT RunLengthTextureFeaturesImageFilter
-  itkConceptMacro( InputPixelTypeCheck,
-                   ( Concept::IsInteger< typename InputImageType::PixelType>) );
+//  itkConceptMacro( InputPixelTypeCheck,
+//                   ( Concept::IsInteger< typename InputImageType::PixelType>) );
```
delete

blowekamp reviewed (commented):
You probably should not convert all tests to floating point input. You should still test with the integer types too.

For the general case just doing a cast from float to int is not sufficient. You need a linear mapping from the range of the floats to the range of the histogram and provide consistent rounding and boundary conditions. This is really where the usefulness of a separate histogram class come into play as it it tricky to implement so that is works for both integers and real types.

jbvimort commented:
Thank you so much for the help! :smiley: 

Concerning the tests I moved about half of them to work with float but obviously kept some with int input type, do you think it would be better to make all tests available for both types of it's fine like that?

I moved everything to work with static cast (largely used across ITK) and a clearer typedef as you suggested. Therefore, the  bounds numbers and floating point errors shouldn't be a problem, indeed the image used in the filter is digitized during the pre-processing step resulting in an image with a range of [-10; binNumber] where binNumber should never exceed 256.

Do you think this new modification is ok?

blowekamp commented:
Should the DigitizedImage's pixel type be an integer? It sounds like this may be the conversion for the continuous pixel space to the discrete histogram index space?

jbvimort commented:
That would make a lot of sense, let me work something out! :smiley:

jbvimort commented:
This should be way better! :smiley:

blowekamp commented:
That looks a lot clearer!

Are you able to work in using a prober rounding method when converting to the HistogramIndexType such as `itk::Math::Floor`?

thewtex commented on example/computeGLCMFeatures.py:
```diff
@@ -6,8 +6,12 @@
 
-im = itk.imread(sys.argv[1])
-mask = itk.imread(sys.argv[2])
+imReader = itk.ImageFileReader.ISS3.New()
```
It is preferred to use:

```
Dimension = 3
PixelType = itk.ctype('signed short')
ImageType = itk.Image[PixelType, Dimension]
imReader = itk.ImageFileReader[ImageType].New()
```
It is longer, but it is more explicit, which is especially important for examples.

thewtex commented on example/computeGLCMFeatures.py:
```diff
@@ -6,8 +6,12 @@
+imReader = itk.ImageFileReader.ISS3.New()
+imReader.SetFileName(sys.argv[1])
+im = imReader.GetOutput()
+maskReader = itk.ImageFileReader.IUC3.New()
```
same here

thewtex commented on example/computeGLRLMFeatures.py:
```diff
@@ -7,14 +7,18 @@
 
-im = itk.imread(sys.argv[1])
-mask = itk.imread(sys.argv[2])
+imReader = itk.ImageFileReader.ISS3.New()
```
explicit

thewtex commented on example/computeGLRLMFeatures.py:
```diff
@@ -7,14 +7,18 @@
+imReader = itk.ImageFileReader.ISS3.New()
+imReader.SetFileName(sys.argv[1])
+im = imReader.GetOutput()
+maskReader = itk.ImageFileReader.IUC3.New()
```
explicit

thewtex commented on include/itkCoocurrenceTextureFeaturesImageFilter.h:
```diff
@@ -103,13 +103,16 @@ class ITK_TEMPLATE_EXPORT CoocurrenceTextureFeaturesImageFilter
 
+
+  typedef int                                                            HistogramIndexType;
+  typedef itk::Image< HistogramIndexType, TInputImage::ImageDimension >  DigitizedImageType;
```
It is preferred to make these typedefs `private` or `protected` so that the value can be changed in the future without worrying about breaking API backwards compatibility.

jbvimort replied:
The problem is that it's used NeighborhoodIteratorType that need to be public to allow to access the NeighborhoodRadius variable. Can it still be either protected or private?

thewtex commented on include/itkRunLengthTextureFeaturesImageFilter.h:
```diff
@@ -116,13 +116,15 @@ class ITK_TEMPLATE_EXPORT RunLengthTextureFeaturesImageFilter
   typedef typename InputImageType::PointType    PointType;
 
+  typedef int                                                            HistogramIndexType;
+  typedef itk::Image< HistogramIndexType, TInputImage::ImageDimension >  DigitizedImageType;
```
protected or private

thewtex reviewed (changes requested):
Nice! A few minor issues noted inline.

jbvimort commented:
@blowekamp a rounding method has been added to the digitalizer in order to make sure that everything is working properly! Thanks for the advice! :smiley:

blowekamp commented:
Looks good to me now. Thanks for addressing my comment.

Making extra typedefs protected, or function scope would be good as Matt requested. I believe a public typedef can reference a private typedef. Give it a try to check.

jbvimort committed:
ENH: Modification of the internal typedef to protected for a better API backword compatibility

jbvimort committed:
ENH: Modification of the version number

jbvimort commented:
The typedef problem has been addressed thanks to Matt advice. :+1: 

Does it look ok for merging?

thewtex reviewed (approved):
Looks great, thanks! :1st_place_medal:
